includes/api/ApiQueryAllUsers.php

```diff
@@ -0,0 +1,383 @@
+		$params = $this->extractRequestParams();
+
+		$prop = $params['prop'];
+		if ( !is_null( $prop ) ) {
+			$prop = array_flip( $prop );
+			$fld_blockinfo = isset( $prop['blockinfo'] );
+			$fld_editcount = isset( $prop['editcount'] );
+			$fld_groups = isset( $prop['groups'] );
+			$fld_rights = isset( $prop['rights'] );
+			$fld_registration = isset( $prop['registration'] );
+			$fld_implicitgroups = isset( $prop['implicitgroups'] );
+		} else {
+			$fld_blockinfo = $fld_editcount = $fld_groups = $fld_registration = $fld_rights = $fld_implicitgroups = false;
+		}
+
+		$limit = $params['limit'];
+
+		$this->addTables( 'user' );
+		$useIndex = true;
+
+		$dir = ( $params['dir'] == 'descending' ? 'older' : 'newer' );
+		$from = is_null( $params['from'] ) ? null : $this->keyToTitle( $params['from'] );
+		$to = is_null( $params['to'] ) ? null : $this->keyToTitle( $params['to'] );
+
+		# MySQL doesn't seem to use 'equality propagation' here, so like the
+		# ActiveUsers special page, we have to use rc_user_text for some cases.
+		$userFieldToSort = $params['activeusers'] ? 'rc_user_text' : 'user_name';
+
+		$this->addWhereRange( $userFieldToSort, $dir, $from, $to );
+
+		if ( !is_null( $params['prefix'] ) ) {
+			$this->addWhere( $userFieldToSort .
+				$db->buildLike( $this->keyToTitle( $params['prefix'] ), $db->anyString() ) );
+		}
+
+		if ( !is_null( $params['rights'] ) ) {
+			$groups = array();
+			foreach( $params['rights'] as $r ) {
+				$groups = array_merge( $groups, User::getGroupsWithPermission( $r ) );
+			}
+
+			$groups = array_unique( $groups );
+
+			if ( is_null( $params['group'] ) ) {
+				$params['group'] = $groups;
+			} else {
+				$params['group'] = array_unique( array_merge( $params['group'], $groups ) );
+			}
+		}
+
+		if ( !is_null( $params['group'] ) && !is_null( $params['excludegroup'] ) ) {
+			$this->dieUsage( 'group and excludegroup cannot be used together', 'group-excludegroup' );
+		}
+
+		if ( !is_null( $params['group'] ) && count( $params['group'] ) ) {
+			$useIndex = false;
+			// Filter only users that belong to a given group
+			$this->addTables( 'user_groups', 'ug1' );
+			$this->addJoinConds( array( 'ug1' => array( 'INNER JOIN', array( 'ug1.ug_user=user_id',
+					'ug1.ug_group' => $params['group'] ) ) ) );
+		}
+
+		if ( !is_null( $params['excludegroup'] ) && count( $params['excludegroup'] ) ) {
+			$useIndex = false;
+			// Filter only users don't belong to a given group
+			$this->addTables( 'user_groups', 'ug1' );
+
+			if ( count( $params['excludegroup'] ) == 1 ) {
+				$exclude = array( 'ug1.ug_group' => $params['excludegroup'][0] );
+			} else {
+				$exclude = array( $db->makeList( array( 'ug1.ug_group' => $params['excludegroup'] ), LIST_OR ) );
+			}
+			$this->addJoinConds( array( 'ug1' => array( 'LEFT OUTER JOIN',
+				array_merge( array( 'ug1.ug_user=user_id' ), $exclude )
+				)
+			) );
+			$this->addWhere( 'ug1.ug_user IS NULL' );
+		}
+
+		if ( $params['witheditsonly'] ) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "witheditsonly" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)